ddl/placement/rule.go

```diff
@@ -0,0 +1,133 @@
+	Leader PeerRoleType = "leader"
+	// Follower matches a follower.
+	Follower PeerRoleType = "follower"
+	// Learner matches a learner.
+	Learner PeerRoleType = "learner"
+)
+
+// Rule is the core placement rule struct. Check https://github.com/tikv/pd/blob/master/server/schedule/placement/rule.go.
+type Rule struct {
+	GroupID        string       `json:"group_id"`
+	ID             string       `json:"id"`
+	Index          int          `json:"index,omitempty"`
+	Override       bool         `json:"override,omitempty"`
+	StartKeyHex    string       `json:"start_key"`
+	EndKeyHex      string       `json:"end_key"`
+	Role           PeerRoleType `json:"role"`
+	Count          int          `json:"count"`
+	Constraints    Constraints  `json:"label_constraints,omitempty"`
+	LocationLabels []string     `json:"location_labels,omitempty"`
+	IsolationLevel string       `json:"isolation_level,omitempty"`
+}
+
+// NewRules constructs []*Rule from a yaml-compatible representation of
+// array or map of constraints. It converts 'CONSTRAINTS' field in RFC
+// docs/design/2021-06-24-placement-rules-in-sql.md to structs.
+func NewRules(replicas uint64, cnstr string) ([]*Rule, error) {
+	rules := []*Rule{}
+
+	cnstbytes := []byte(cnstr)
+
+	constraints1 := []string{}
+	err1 := yaml.UnmarshalStrict(cnstbytes, &constraints1)
+	if err1 == nil {
+		// can not emit REPLICAS with an array or empty label
+		if replicas == 0 {
+			return rules, fmt.Errorf("%w: should be positive", ErrInvalidConstraintsRelicas)
+		}
+
+		labelConstraints, err := NewConstraints(constraints1)
+		if err != nil {
+			return rules, err
+		}
+
+		rules = append(rules, &Rule{
+			Count:       int(replicas),
+			Constraints: labelConstraints,
+		})
+
+		return rules, nil
+	}
+
+	constraints2 := map[string]int{}
+	err2 := yaml.UnmarshalStrict(cnstbytes, &constraints2)
+	if err2 == nil {
+		ruleCnt := 0
+		for labels, cnt := range constraints2 {
+			if cnt <= 0 {
+				return rules, fmt.Errorf("%w: count of labels '%s' should be positive, but got %d", ErrInvalidConstraintsMapcnt, labels, cnt)
+			}
+			ruleCnt += cnt
+		}
+
+		if replicas == 0 {
+			replicas = uint64(ruleCnt)
+		}
+
+		if int(replicas) < ruleCnt {
+			return rules, fmt.Errorf("%w: should be larger or equal to the number of total replicas, but REPLICAS=%d < total=%d", ErrInvalidConstraintsRelicas, replicas, ruleCnt)
+		}
+
+		for labels, cnt := range constraints2 {
+			labelConstraints, err := NewConstraints(strings.Split(labels, ","))
+			if err != nil {
+				return rules, err
+			}
+
+			rules = append(rules, &Rule{
+				Count:       cnt,
+				Constraints: labelConstraints,
+			})
+		}
+
+		remain := int(replicas) - ruleCnt
+		if remain > 0 {
+			rules = append(rules, &Rule{
+				Count: remain,
+			})
+		}
+
+		return rules, nil
+	}
+
+	return nil, multierror.Append(fmt.Errorf("%w: should be [constraint1, ...], {constraint1: cnt1, ...}, or any yaml compatible representation", ErrInvalidConstraintsFormat), err1, err2)
```

Comment:
Why do we need `multierror`? `NewRule` can be split into 2 functions: `NewRuleFromArr` and `NewRuleFromMap`, then a simple nested `if` statement is enough.

Reply:
Multierror was used for other purposes. So here it is used for possible inspection, in favor of `%s`, which stringify the error into a plain string.